LayoutBuilder template using either ForthMachine32 or ForthMachine64 and a tutorial (#1063)
jpivarski/awkward-1.0: docs-src/how-to-create-layoutbuilder.md

How to create arrays with LayoutBuilder (more control)
======================================================

What is LayoutBuilder
---------------------

`ak.layout.LayoutBuilder` is the low-level LayoutBuilder that builds layouts, or `ak.layout.Content` arrays. It must be initialized by a JSON string that represents a valid `ak.forms.Form`.

> A layout is the composable `ak.layout.Content` elements that determine how an array is structured. The layout may be considered a "low-level" view, as it distinguishes between arrays that have the same logical meaning (i.e. same JSON output and high-level `type`) but different

Reviewer: Maybe clearer to say
```suggestion
A layout is build from composable `ak.layout.Content` elements that determine how an array is structured. The layout may be considered a "low-level" view, as it distinguishes between arrays that have the same logical meaning (i.e. same JSON output and high-level `type`) but different
```

Reviewer 2: I think "is" is correct: a layout is the tree of nested `ak.layout.Content` nodes. To say that it's built from them implies that there's another step, analogous to the step from a Builder (ArrayBuilder or LayoutBuilder) to an array through `snapshot`.

"Layout" may simply be a synonym for `Content`, though it may be distinguished in that "layout" often refers to the whole tree and a `Content` object is specifically one node, since it has a meaning in code. Maybe that's what you meant by "built from."

In that case, how about "A layout consists of composable `ak.layout.Content`..."? The word "consists" is describing a static thing, whereas "built from" could describe a static thing or could describe a process like the one involving `snapshot`.

Reviewer: That is better still (and doesn't have a typo)!

Author: Thanks! It's fixed.

  * node types, such as `ak.layout.ListArray64` and `ak.layout.ListOffsetArray64`,
  * integer type specialization, such as `ak.layout.ListArray64` and `ak.layout.ListArray32`,
  * or specific values, such as gaps in a `ak.layout.ListArray64`.

> `ak.forms.Form` describes a low-level data type or "form". There is an exact one-to-one relationship between each `ak.layout.Content` class and each Form.

Reviewer: Maybe lowercase form, or link to `ak.forms.Form` again?
```suggestion
`ak.forms.Form` describes a low-level data type or "form". There is an exact one-to-one relationship between each `ak.layout.Content` class and each form.
```

Reviewer 2: I'd prefer capitals on "Form" because it's such a common word and easily mis-seen or even mistyped as "from." I try to use capitals with "Form" because it clues people in to "this is some special thing with a specific meaning."

All of these references to classes could point to the documentation: instead of "Form" or even "`ak.forms.Form`", you could say [ak.forms.Form](https://awkward-array.readthedocs.io/en/latest/ak.forms.Form.html). Similarly with their subclasses, including Content subclasses like [ak.layout.ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html). I think the other tutorials do this.

```

and continue:

```{code-cell}
builder.begin_list()
builder.float64(4.4)
builder.float64(5.5)
builder.end_list()
```
Remember, you can taka a snapshot at any time:

```{code-cell}
layout = builder.snapshot()
layout
```

Nested records
--------------

When using a RecordArray form you do not have to specify a field, the fields alternate.

```{code-cell}
form = """
{
"class": "RecordArray",
"contents": {
    "one": "float64",
    "two": "int64"
},
"form_key": "node0"
}
"""
builder = ak.layout.LayoutBuilder32(form)

# the fields alternate
builder.float64(1.1)  # "one"
builder.int64(2)      # "two"
builder.float64(3.3)  # "one"
builder.int64(4)      # "two"
```

```{code-cell}
layout = builder.snapshot()
layout
```

Similarly, for the record contents with the same type:

```{code-cell}
form = """
{
"class": "RecordArray",
"contents": {
    "one": "float64",
    "two": "float64"
},
"form_key": "node0"
}
"""
```

```{code-cell}
builder = ak.layout.LayoutBuilder32(form)
```

If record contents have the same type, the fields alternate:

```{code-cell}
builder.float64(1.1)  # "one"
builder.float64(2.2)  # "two"
builder.float64(3.3)  # "one"
builder.float64(4.4)  # "two"

> ```

Reviewer: Can we still specify the field if we want? If so, could we mention that?

````suggestion
```
However, the fields can be explicitly specified if required:
``{code-cell}
builder.field("two").float64(1.1)  # "two"
builder.field("one").float64(2.2)  # "one"
builder.field("one").float64(3.3)  # "one"
builder.field("two").float64(4.4)  # "two"
```
````

Reviewer 2: I don't remember whether `builder.field("xyz")` acts as an assertion or not. It will not allow you to write them in a different order: that's a performance thing.

Author: No, there is no command to set a field. I've changed the wording to make it more clear. Thanks!

Reviewer: Ah, good to know!

  "index": "i64",
  "contents": [
      "float64",
      "bool",
      "int64"
  ],
  "form_key": "node0"
}
  """
```

When a builder is created, it will accept only "float64", "bool", and "int64" data. The builder data methods to append these types have similar to the types names: `float64`, `boolean`, and `int64`.

```{code-cell}
builder = ak.layout.LayoutBuilder(form)
```

Each of the union array contents has a tag associated with it. The tags are contiguous integers starting with `0` for the first content. You have to issue a `tag` command prior to each data method. The order of filling the contents can be arbitrary - the `tag` command uniquely identifies the content to which the data is appended:

```{code-cell}
builder.tag(0)
builder.float64(1.1)
builder.tag(1)
builder.boolean(False)
builder.tag(2)
builder.int64(11)
builder.tag(0)
builder.float64(2.2)
builder.tag(1)
builder.boolean(False)
builder.tag(0)
builder.float64(2.2)
builder.tag(0)
builder.float64(3.3)
builder.tag(1)
builder.boolean(True)
builder.tag(0)
builder.float64(4.4)
builder.tag(1)
builder.boolean(False)
builder.tag(1)
builder.boolean(True)
builder.tag(0)
builder.float64(-2.2)
```

Snapshot
--------

> To turn a #ak::LauoutBuilder into a layout, call `snapshot`. This is an inexpensive operation (may be done multiple times; the builder is unaffacted).

Reviewer: ```suggestion
To turn a #ak::LayoutBuilder into a layout, call `snapshot`. This is an inexpensive operation (may be done multiple times; the builder is unaffected).
```

Reviewer: I took a look over this out of curiosity - let me know if you'd rather wait for review before any suggestions. I can't do much besides code review at the moment so I'm keen to help!

Author: thanks, I'll pin you when I'm finished!

Author: @agoose77 - Can I take you up on your kind offer of code review? Thanks!

Author: @agoose77 - Thanks for the review - very much appreciated
